Add websockets surface module

lucianoiam opened the pull request:
This is a new surface module that implements the WebSockets protocol. The overall architecture was inspired by OSC, client reads and writes values at certain locations. In a nutshell:

`node (optional_address) = value`

Payload is JSON based, for example the following sets plugin parameter with id=2 at plugin id=1 in strip id=0 to value 66.6:

`{"node":"strip_plugin_param_value","addr”:[0,1,2],”val”:[66.6]}`

**Currently supported features**

- Session tempo
- Strip gain, pan and mute controls
- Strip meter feedback
- Plugin bypass
- Plugin parameter control

**Pending work**

- UI for configuring TCP port (now hardcoded to 9000)
- Documentation

PR includes a web browser based demo client. Surface module requires libwebsockets >= 2. Tested on Ubuntu and macOS. Should work on Windows (not tested yet).

x42 commented:
Nice! That looks like a good approach. +1 for using libwebsockets.

Alas, it currently fails to compile here, due to _C++11isms_.  Try to configure with `./waf configure --strict ...` that adds `--ansi` (C++98 compatibility):

Compile log: http://paste.debian.net/1131273/

might be an easy fix or minor refactoring, and some `boost::unordered_map` instead of `std::unordered_map`.

lucianoiam commented:
Thanks for reviewing. I pushed an update that swaps `std::unordered_set` and `std::unordered_map` for their boost equivalents. Removed some `std::stoi` and similar type conversions unavailable in older C++ standards in favor of `boost::lexical_cast`. Also removed some non-static members initialization in headers and tweaked a couple more details.

The code still contains at least one C++11 feature namely using initializer lists for vectors; those can also be removed in detriment of code clarity. I fear not being able to get rid of all non C++98 features without creating the right dev environment first, my Ubuntu box has all off-the-shelf dev packages installed and seems several of them are C++11 only (for example the glibmm headers which according to their git history went C++11 around 2015). Ardour just won’t build under such environment when passing `--strict`. 

( Strangely enough `--strict` already worked on Mac for the original PR, maybe it is because of clang instead of gcc? )

I am not experienced enough and missed a lot of history on C++ to rely on my visual code inspection skills only; there could be more details that should be took care of. I guess a late comer to C++ needs the compiler to shout out as many errors as possible.

How could I recreate the nightly builds development environment? is there a list of exact versions for each dependency? I could try setting up such an environment to test compilation before pushing further updates.

x42 commented:
>  ( Strangely enough --strict already worked on Mac for the original PR, maybe it is because of clang instead of gcc? )

The linked build log is from a Mac (OSX 10.6, gcc-4.2).

lucianoiam commented:
Here compilation succeeds passing --strict on OSX 10.15 with the standard toolchain installed by latest Xcode (clang based), a different environment compared to yours. I guess I need to setup a new dev machine.

My configure line:

```
luciano@Lucianos-MBP src % ./waf configure --strict --ptformat --with-backends=jack,coreaudio,dummy --optimize
Setting top to                           : /Users/luciano/src/ardour 
Setting out to                           : /Users/luciano/src/ardour/build 
Checking for 'gcc' (C compiler)          : not found 
Checking for 'clang' (C compiler)        : /usr/bin/clang 
Checking for 'g++' (C++ compiler)        : not found 
Checking for 'clang++' (C++ compiler)    : /usr/bin/clang++ 

Global Configuration 
 * Install prefix                                    : /usr/local 
 * Debuggable build                                  : False 
 * Build documentation                               : False 

Ardour Configuration 
 * Will build against private GTK dependency stack   : no 
 * Will use explicit linkage against libintl in /Users/luciano/gtk/inst : yes 
 * Will build against private Ardour dependency stack : no 
No Carbon support available for this build

Checking for function 'dlopen' in dlfcn.h            : yes 
Checking for boost library >= 1.56                   : ok 
Checking for program 'pkg-config'                    : /usr/local/bin/pkg-config 
Checking for 'glib-2.0' >= 2.28                      : yes 
Checking for 'gthread-2.0' >= 2.2                    : yes 
Checking for 'glibmm-2.4' >= 2.32.0                  : yes 
Checking for 'sndfile' >= 1.0.18                     : yes 
Checking for 'giomm-2.4' >= 2.2                      : yes 
Checking for 'libcurl' >= 7.0.0                      : yes 
Checking for 'libarchive' >= 3.0.0                   : yes 
Checking for 'liblo' >= 0.26                         : yes 
Checking for 'taglib' >= 1.6                         : yes 
Checking for 'vamp-sdk' >= 2.1                       : yes 
Checking for 'vamp-hostsdk' >= 2.1                   : yes 
Checking for 'rubberband'                            : yes 
Checking for  sndfile RF64=>RIFF support             : Found 
Checking for clang                                   : yes 
Checking for compiler flags ['-std=c++11']           : yes 

Warning: you are building Ardour with SSE support even though your system does not support these instructions. (This may not be an error, especially if you are a package maintainer)
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'fftw3f'                                : yes 
Checking for 'aubio' >= 0.3.2                        : yes 
Checking for 'aubio' >= 0.4.0                        : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'libxml-2.0'                            : yes 
Checking for 'sigc++-2.0' >= 2.0                     : yes 
Checking for header execinfo.h                       : yes 
Checking for header unistd.h                         : yes 
Checking for function 'posix_memalign' in stdlib.h   : yes 
Checking for function 'getmntent' in mntent.h        : no 
Checking for function 'localtime_r' in time.h        : yes 
Checking for header boost/shared_ptr.hpp             : yes 
Checking for header boost/weak_ptr.hpp               : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'cppunit' >= 1.12.0                     : yes 
Checking for header boost/shared_ptr.hpp             : yes 
Checking for header boost/weak_ptr.hpp               : yes 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for header boost/shared_ptr.hpp             : yes 
Checking for header boost/weak_ptr.hpp               : yes 
Checking for 'libusb-1.0'                            : yes 
Checking for header cwiid.h                          : not found 
You are missing the cwiid headers needed to compile wiimote support
Checking for 'libwebsockets' >= 2.0.0                : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'pangomm-1.4' >= 1.4                    : yes 
Checking for 'cairomm-1.0' >= 1.8.4                  : yes 
Checking for 'jack' >= 0.121.0                       : yes 
Checking for JACK metadata API                       : ok 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for program 'gas, gcc'                      : /usr/bin/gcc 
Checking for program 'ar'                            : /usr/bin/ar 
Checking for 'lrdf' >= 0.4.0                         : yes 
Checking for 'samplerate' >= 0.1.0                   : yes 
Checking for 'lv2' >= 1.2.0                          : yes 
Checking for 'lv2' >= 1.10.0                         : yes 
Checking for 'serd-0' >= 0.14.0                      : yes 
Checking for 'sord-0' >= 0.8.0                       : yes 
Checking for 'sratom-0' >= 0.2.0                     : yes 
Checking for 'lilv-0' >= 0.24.2                      : yes 
Checking for 'suil-0' >= 0.6.0                       : yes 
Checking for 'ogg' >= 1.1.2                          : yes 
Checking for 'flac' >= 1.2.1                         : yes 
Checking for 'fftw3f' >= 3.3.5                       : yes 
Checking for header sys/vfs.h                        : not found 
Checking for header sys/statvfs.h                    : yes 
Checking for header unistd.h                         : yes 
Checking for header boost/shared_ptr.hpp             : yes 
Checking for header boost/weak_ptr.hpp               : yes 
Checking for header boost/scoped_ptr.hpp             : yes 
Checking for header boost/ptr_container/ptr_list.hpp : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'gtkmm-2.4' >= 2.8                      : yes 
Checking for 'gtk+-2.0' >= 2.12.1                    : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'samplerate' >= 0.1.7                   : yes 
Checking for header boost/shared_ptr.hpp             : yes 
Checking for header boost/format.hpp                 : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'lv2' >= 1.0.0                          : yes 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'cairo' >= 1.12.0                       : yes 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'gthread-2.0' >= 2.10.1                 : yes 
Checking for 'gtk+-2.0' >= 2.18                      : yes 
Checking for 'x11' >= 1.1                            : not found 
Checking for 'pangoft2' >= 1.36.8                    : yes 
Checking for 'fontconfig'                            : yes 
Checking for Cocoa live resizing constants           : yes 
Checking for header boost/shared_ptr.hpp             : yes 
Checking for header boost/weak_ptr.hpp               : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
 * build session-utils                               : yes 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for 'gcc' (C compiler)                      : not found 
Checking for 'clang' (C compiler)                    : /usr/bin/clang 
Checking for 'g++' (C++ compiler)                    : not found 
Checking for 'clang++' (C++ compiler)                : /usr/bin/clang++ 
Checking for header stdio.h readline/readline.h      : yes 
 * Build documentation                               : False 
 * Debuggable build                                  : False 
 * Export all symbols (backtrace)                    : False 
 * Install prefix                                    : /usr/local 
 * Strict compiler flags                             : [] 
 * Internal Shared Libraries                         : True 
 * Use External Libraries                            : False 
 * Library exports hidden                            : True 
 * Free/Demo copy                                    : False 
 * ALSA DBus Reservation                             : False 
 * Architecture flags                                : None 
 * Aubio                                             : True 
 * AudioUnits                                        : True 
 * Build target                                      : catalina 
 * Canvas Test UI                                    : False 
 * Beatbox test app                                  : False 
 * CoreAudio                                         : True 
 * CoreAudio 10.5 compat                             : False 
 * Debug RT allocations                              : False 
 * Debug Symbols                                     : False 
 * Denormal exceptions                               : False 
 * FLAC                                              : True 
 * FPU optimization                                  : True 
 * Freedesktop files                                 : False 
 * Libjack linking                                   : weak 
 * Libjack metadata                                  : True 
 * Lua Binding Doc                                   : False 
 * Lua Commandline Tool                              : True 
 * LV2 UI embedding                                  : True 
 * LV2 support                                       : True 
 * LV2 extensions                                    : True 
 * LXVST support                                     : False 
 * Mac VST support                                   : True 
 * NI-Maschine                                       : False 
 * OGG                                               : True 
 * Phone home                                        : True 
 * Process thread timing                             : False 
 * Program name                                      : Ardour 
 * Samplerate                                        : True 
 * PT format                                         : True 
 * PTW32 Semaphore                                   : False 
 * Threaded WaveViews                                : True 
 * Translation                                       : True 
 * Unit tests                                        : False 
 * Use LLD linker                                    : False 
 * Windows VST support                               : False 
 * Wiimote support                                   : False 
 * Windows key                                       : Mod4><Super 
 * PortAudio Backend                                 : False 
 * CoreAudio/Midi Backend                            : True 
 * ALSA Backend                                      : False 
 * Dummy backend                                     : True 
 * JACK Backend                                      : True 
 * Pulseaudio Backend                                : False 
 * Buildstack                                        : -system- 
 * Mac i386 Architecture                             : False 
 * Mac ppc Architecture                              : False 
 * C compiler flags                                  : ['-I/Users/luciano/src/ardour', '-DHAVE_RF64_RIFF', '-DWAF_BUILD', '-DNDEBUG', '-std=c99', '-pedantic', '-Wshadow', '-Wall', '-Wcast-align', '-Wextra', '-Wwrite-strings', '-fshow-column', '-O3', '-ffast-math', '-pipe', '-DARCH_X86', '-DBUILD_VECLIB_OPTIMIZATIONS', '-DMAC_OS_X_VERSION_MAX_ALLOWED=1090', '-mmacosx-version-min=10.9', '-Wall', '-Wpointer-arith', '-Wcast-qual', '-Wcast-align', '-Wno-unused-parameter', '-DBOOST_SYSTEM_NO_DEPRECATED', '-D_ISOC9X_SOURCE', '-D_LARGEFILE64_SOURCE', '-D_FILE_OFFSET_BITS=64', '-DPROGRAM_NAME="Ardour"', '-DPROGRAM_VERSION="6"', '-Qunused-arguments', '-Wstrict-prototypes', '-Wmissing-prototypes'] 
 * C++ compiler flags                                : ['-I/Users/luciano/src/ardour', '-DHAVE_RF64_RIFF', '-DWAF_BUILD', '-DNDEBUG', '-ansi', '-Wnon-virtual-dtor', '-Woverloaded-virtual', '-Wall', '-Wcast-align', '-Wextra', '-Wwrite-strings', '-fshow-column', '-O3', '-ffast-math', '-pipe', '-DARCH_X86', '-DBUILD_VECLIB_OPTIMIZATIONS', '-DMAC_OS_X_VERSION_MAX_ALLOWED=1090', '-mmacosx-version-min=10.9', '-Wall', '-Wpointer-arith', '-Wcast-qual', '-Wcast-align', '-Wno-unused-parameter', '-DBOOST_SYSTEM_NO_DEPRECATED', '-D_ISOC9X_SOURCE', '-D_LARGEFILE64_SOURCE', '-D_FILE_OFFSET_BITS=64', '-DPROGRAM_NAME="Ardour"', '-DPROGRAM_VERSION="6"', '--stdlib=libc++', '-std=c++11', '-DCARBON_FLAT_HEADERS', '-Qunused-arguments', '-Woverloaded-virtual', '-Wno-mismatched-tags', '-Wno-cast-align', '-Wno-unused-local-typedefs', '-Wunneeded-internal-declaration', '-D__STDC_LIMIT_MACROS', '-D__STDC_FORMAT_MACROS', '-DCANVAS_COMPATIBILITY', '-DCANVAS_DEBUG'] 
 * Linker flags                                      : ['-Xlinker', '-headerpad_max_install_names', '--stdlib=libc++'] 

'configure' finished successfully (9.380s)
```

lucianoiam commented:
I think the code is C++98 complaint now, it compiles passing --strict to gcc.

x42 commented:
Yay! Thanks for all the hard work! It does indeed compile on all relevant systems now.
The source reads nicely, too.

I'm tempted to merge it in time for tomorrow's nighty build.
@pauldavisthefirst any last word?

x42 commented:
I had to prefix `POLLIN` `POLLOUT` etc with `LWS_` to fix windows builds.
LWS doesn't define POLLERR.  `include/libwebsockets.h` on unix systems combines 

` LWS_POLLHUP (POLLHUP|POLLERR)`

This fixes Ardour/Windows builds (untested):

```diff
--- a/libs/surfaces/websockets/server.cc  2020-02-22 19:01:15.989652876 +0100
+++ b/libs/surfaces/websockets/server.cc  2020-02-22 19:02:55.209734016 +0100
@@ -163,7 +163,7 @@
 
     it->second.lws_pfd.events = pa->events;
 
-    if (pa->events & POLLOUT) {
+    if (pa->events & LWS_POLLOUT) {
         // libwebsockets wants to write but cannot find a way to update
         // an existing glib::iosource event flags using glibmm,
         // create another iosource and set to IO_OUT, it will be destroyed
@@ -303,15 +303,15 @@
 {
     IOCondition ioc = Glib::IO_ERR;
 
-    if (events & POLLIN) {
+    if (events & LWS_POLLIN) {
         ioc |= Glib::IO_IN;
     }
 
-    if (events & POLLOUT) {
+    if (events & LWS_POLLOUT) {
         ioc |= Glib::IO_OUT;
     }
 
-    if (events & POLLHUP) {
+    if (events & LWS_POLLHUP) {
         ioc |= Glib::IO_HUP;
     }
 
@@ -324,19 +324,23 @@
     int events = 0;
 
     if (ioc & Glib::IO_IN) {
-        events |= POLLIN;
+        events |= LWS_POLLIN;
     }
 
     if (ioc & Glib::IO_OUT) {
-        events |= POLLOUT;
+        events |= LWS_POLLOUT;
     }
 
     if (ioc & Glib::IO_HUP) {
-        events |= POLLHUP;
+        events |= LWS_POLLHUP;
     }
 
     if (ioc & Glib::IO_ERR) {
-        events |= POLLERR;
+#ifdef PLATFORM_WINDOWS
+        events |= LWS_POLLHUP;
+#else
+        events |= LWS_POLLERR;
+#endif
     }

```

lucianoiam commented:
Thanks! I learn quite a few things doing this and this module is actually useful for my setup.
I plan to write some client stuff, though I still have to decide on the approach.

Great to know is also compiling on Windows; I pushed an update that adds the `LWS_` prefixes and also slightly changes how the `event_to_ioc()` method works.  The version I posted originally started with `ioc=IO_ERR` then or’ed any additional flags. Since the `io_handler()` does not check for `IO_ERR` later this caused no harm but I think the original version was conceptually wrong.

In regards to this topic, seems libwesockets >= 3 already includes support for integrating into a glib loop and I expect it to be a better implementation that mine’s; but decided to keep v2 compatibility and roll my own solution because Ubuntu Studio repos are stuck at lws 2. I guess it is a popular Linux distro among music makers out there so being able to just apt-get libwebsockets seemed worth the effort.

x42 commented:
We use  libwebsockets 3.0.99 (see https://nightly.ardour.org/list.php#build_deps) for Ardour binaries on all platforms. - I don't know about UbStu, but KXStudio and AVLinux ship binaries from ardour.org and do not rely on Ubuntu dependencies.

What remains need to happen for this ctrl surface to loose "Experimental" as part of the user visible name?

lucianoiam commented:
Thanks for that link that helps tuning my dev environment.

Well I guess the label is there because this started as a personal experiment but after a few iterations it grew stable enough and Experimental remained; I think it can be removed but I'd leave this up to the Ardour team to decide. Haven't run any formal tests but see no crashes and the feature set though reduced it works as expected.

x42 commented:
Rebased and merged as 6.0-pre0-3473 d7ef188aea

Thank you!

x42 commented:
I see now, you mentioned websocket 2.x because those `LWS_POLL*` defines are new in 3.x?

lucianoiam commented:
It seems so, just realized when trying to recompile on my Linux (which has 2.x installed). But the last patch you committed fixes that.
